### ISSUE TITLE
ML Efficiency RMSE results not comparable across methods

### ISSUE THREAD

**Issue by Ajoo** [muellermarkus/cdtd#2]
Hi @muellermarkus. 

I really like the evaluation setup form your paper and find that it fixes many of the issues I have with setups from other papers on generative models for tabular data. Over the past few days I've been trying to add a few other recent methods to your benchmark, starting with TabDiff, but I quickly ran into an issue when I started looking at the ML efficiency metrics for regression. Take for example the following table for the `nmes` dataset where catboost trained on real data is better than generated data when looking at $R^2$ but the opposite appears to be true when looking at RMSE:

|     model      |  rmse |   r2   |
|----------------|-------|--------|
|  real_forest   | 1.029 | 0.102  |
|  real_linear   | 0.996 | 0.159  |
| real_catboost  | 0.977 |  0.19  |
|  synth_forest  | 0.472 | -1.388 |
|  synth_linear  | 0.388 | -0.613 |
| synth_catboost | 0.383 | -0.572 |

I was puzzled by this since R^2 is just a normalized version of rmse (`R2_test = 1 - rmse_test^2/var(y_test)`). After digging a bit into the issue I realized that the problem is that the target is being rescaled by the std of the training data: https://github.com/muellermarkus/cdtd/blob/5bfab87cb7f4b364d6a1b43ec62ac9efc4e667bd/evaluation/eval_ml_efficiency.py#L234-L241

This scaling factor therefore depends on the generated training data itself which renders rmse results not comparable across different methods (or even different synthetic sets generated by the same method). In other words, the reported rmse for synthetic data is not comparable to the number reported for real data, nor to rmse of other synthetic methods. Note that this scaling benefits methods that generate target variable data with larger variance as this will scale the rmse, as measured in the original test targets, down by a larger factor. I suspect that this is the reason synthetic data appears to outperform real data on the `news` and `nmes` datasets for many of the methods.

Note that the computed R^2 results, however, should be comparable since this metric is invariant to an affine transformation of `y_test`, but these are not reported in the paper. Personally I am biased to always prefer looking at R^2 anyways, since it is naturally adimensional and immediately tells me if a regressor is doing better than a constant prediction without the need to look at a second number...

Best regards

**Comment by muellermarkus** [comment]
Hi @Ajoo ,

nice catch! Thanks a lot for bringing this to my attention!

Indeed, using the generated data for standardization as well makes the absolute metrics less comparable across generative models. If you care about the RMSE values themselves, this of course matters. Note that if you care about only evaluating the generative performance of a model, this is less important. We define machine learning efficiency as the absolute difference between train set RMSE and generated data RMSE. Hence, if the generative model gets the variance wrong, this simply adds to the absolute difference by scaling the generated data RMSE up/down. Since the train set RMSE is the same for all generative models on a given dataset, a generative model that generates targets with a much larger variance is therefore penalized. Even though the MLE evaluation is not perfect, only caring about absolute differences alleviates the issue somewhat. Let me know if you agree with this point!

I also agree that the R-squared metric could be preferable here. However, this comes back to a fundamental problem on evaluating generative models for tabular data: there is no set of standard evaluation metrics. Reviewers will always ask for more metrics but at the same time reporting every metric is also infeasible. The issue with MLE is even more complicated as the metric depends on which models you are using to estimate.

Best,
Markus

**Comment by Ajoo** [comment]
Hi @muellermarkus, thanks for the reply. To be honest, I'm not a big fan of the original motivation for the MLE metric as I personally find it to represent a somewhat convoluted scenario. I prefer to interpret it as a metric for how well a generative model (learning a general $p(x,y)$ ) learns a particular conditional distribution, $p(y\vert x)$. This conditional is what the downstream model has to learn from the generated data and therefore is what matters most when computing this metric. This is why I generally care about the value of the metric itself but I realize this may go against the general consensus and I also get your point.

Regarding metrics, I agree with you, reporting R^2 or RMSE is equivalent to me since you can compute one from the other as long as you know the variance of the test data. I only mentioned it since these are numbers that you probably already have computed.

In any case, I reran the experiments for the diffusion variants because there were other small adjustments I wanted to make to the evaluation. The results I got actually only strengthen your claim since your model now comes up on top in MLE for both news and nmes. I also don't get any results where synthetic data appears better than real data (the natural order of things) so defining MLE as the absolute difference to real data or going by the metric value itself leads to the same conclusions...

As a sidenote, CDTD also seems to do better than TabDiff in my experiments and I believe that, if anything, I was generous to give it more training steps than you setup generally allows.

Best,
João

**Comment by muellermarkus** [comment]
Hi @Ajoo, 

indeed, MLE is a strange (but unfortunately popular) metric. It is very limited in what it conveys. In particular, if many features are not really informative about $p(y|x)$, the associated feature-specific sample qualities are not accounted for at all by MLE. I do see why you would care about the value of the metric itself in your case.

Thanks for stating the results of your additional experiments! It is great that the intuition of synthetic data leading to no better results than the original data is confirmed. 

Regarding the performance relative to TabDiff, I have been informed about experiments conducted by other authors that support a similar claim. I ran some experiments myself but these will be part of a separate paper. While TabDiff generally works well, I also found that its efficiency is severely lacking. This holds in comparison to most other diffusion models for tabular data out there, not just relative to CDTD.

Best,
Markus

**Comment by Ajoo** [comment]
Hi @muellermarkus,

I think the problem with MLE is not even just related to having $x$ variables not predictive of $y$. If you factorize the DGP as $p(x, y) = p(y \vert x) p(x)$, the $p(x)$ part plays only a small part in the MLE metric (increasing the statistical efficiency of the estimator of $p(y|x)$ ). There is a simple experiment that shows this:
- Train a discriminative model to learn $p(y|x)$ (like catboost) on real data
- Generate synthetic data by sampling $x$ from a simple product of marginals (destroying any relationship between $x$ variables) and $y|x$ from the discriminative model
- Evaluate MLE

This dumb baseline is "competitive" on all datasets, even outperforming all other proper generative models on a few of them. This despite the generated $x$ samples being generally very different from real data...

Regarding TabDiff, it actually doesn't do so bad in the detection score metric. That is one of the metrics I like the most since it's the only one actually measuring something like a distance between multivariate probability distributions. But I do have some concerns with the TabSyn eval setup that seems to be getting adopted as somewhat of a standard benchmark in this field. For the most part, it only focuses on discrepancies between generated and training data. I think this is a bad idea because of overfitting, providing a trivial path to "SOTA" results. Just train a model with higher capacity and/or train it for longer... 

Looking forward to your next paper.

Best,
João

PS: feel free to close this issue
